File: core/src/no/elg/infiniteBootleg/world/blocks/TickingBlock.java
package no.elg.infiniteBootleg.world.blocks;

import no.elg.infiniteBootleg.Ticking;
import no.elg.infiniteBootleg.world.Block;
import no.elg.infiniteBootleg.world.Chunk;
import no.elg.infiniteBootleg.world.Material;
import no.elg.infiniteBootleg.world.World;
import org.jetbrains.annotations.NotNull;

/**
 * Describes a block that implements the {@link Ticking} interface.
 * <p>
 * This implementation is set up to automatically call {@link #tick()} and {@link #tickRare()} if {@link #shouldTick()}
 * is {@code true}. To change this behaviour override {@link #shouldTick()}
 * <p>
 * For each chunk has a list of blocks that extends this class and will call {@link #tryTick(boolean)} every tick.
 *
 * @see Chunk
 * @see TntBlock TntBlock (for change of ticking behaviour)
 */
public abstract class TickingBlock extends Block implements Ticking {

    private boolean shouldTick;

    public TickingBlock(@NotNull World world, Chunk chunk, int localX, int localY, @NotNull Material material) {
        super(world, chunk, localX, localY, material);
    }

    /**
     * Update if the update flag is set to true
     *
     * @param rare
     *     If the the rare update should be called instead of the normal update
     */
    public void tryTick(boolean rare) {
        if (shouldTick()) {
            shouldTick = false;
            if (rare) {
                tickRare();
            }
            else {
                tick();
            }
        }
    }

    public boolean shouldTick() {
        return shouldTick;
    }

    public void setShouldTick(boolean shouldTick) {
        this.shouldTick = shouldTick;
    }
}

File: core/src/main/java/no/elg/infiniteBootleg/world/blocks/TickingBlock.kt
package no.elg.infiniteBootleg.world.blocks

import no.elg.infiniteBootleg.world.BlockImpl
import no.elg.infiniteBootleg.world.Chunk
import no.elg.infiniteBootleg.world.Material
import no.elg.infiniteBootleg.world.World
import no.elg.infiniteBootleg.world.blocks.traits.TickingTrait
import no.elg.infiniteBootleg.world.blocks.traits.TickingTraitHandler

/**
 * Describes a block that implements the [Ticking] interface.
 *
 *
 * This implementation is set up to automatically call [.tick] and [.tickRare] if
 * [.shouldTick] is `true`. To change this behaviour override [.shouldTick]
 *
 *
 * For each chunk has a list of blocks that extends this class and will call [ ][.tryTick] every tick.
 *
 * @see Chunk
 *
 * @see TntBlock TntBlock
 */
abstract class TickingBlock protected constructor(
  world: World,
  chunk: Chunk?,
  localX: Int,
  localY: Int,
  material: Material
) : BlockImpl(world, chunk!!, localX, localY, material), TickingTrait {

  private val tickingTrait: TickingTraitHandler

  init {
    tickingTrait = TickingTraitHandler(this, world.worldTicker)
  }

  override fun tryTick(rare: Boolean) {
    tickingTrait.tryTick(rare)
  }

  override fun shouldTick(): Boolean = tickingTrait.shouldTick

  override fun enableTick() {
    tickingTrait.shouldTick = true
  }

  override fun delayedShouldTick(delayTicks: Long) {
    tickingTrait.delayedShouldTick(delayTicks)
  }
}
